addons/survey/models/survey_survey.py

```diff
@@ -427,3 +509,56 @@ def action_survey_user_input_invite(self):
                     'search_default_invite': 1})
         action['context'] = ctx
         return action
+
+    @api.multi
+    def has_attempts_left(self, partner, email, attempts_pool_token):
+        self.ensure_one()
+
+        if (self.access_mode != 'public' or self.users_login_required) and self.is_attempts_limited:
+            return self.attempts_left(partner, email, attempts_pool_token) > 0
+
+        return True
+
+    @api.multi
+    def attempts_left(self, partner, email, attempts_pool_token):
+        """ Returns the number of attempts left. """
+        self.ensure_one()
+
+        domain = [
+            ('survey_id', '=', self.id),
+            ('test_entry', '=', False),
+            ('state', '=', 'done')
+        ]
+
+        if partner:
+            domain = expression.AND([domain, [('partner_id', '=', partner.id)]])
+        else:
+            domain = expression.AND([domain, [('email', '=', email)]])
+
+        if attempts_pool_token:
+            domain = expression.AND([domain, [('attempts_pool_token', '=', attempts_pool_token)]])
+
+        return self.attempts_limit - self.env['survey.user_input'].search_count(domain)
+
+    @api.multi
+    def _generate_question_ids(self):
```

Comment:
def _prepare_answer_questions ? or something like that. Here we could understand that you generate questions.

Reply:
Done.